File: continuum-js/continuum-idl-js/src/api/ArgumentDefinition.ts
import {AbstractDefinition} from '@/api/AbstractDefinition'
import {C3Type} from '@/api/C3Type'
import {C3Decorator} from '@/api/decorators/C3Decorator'

/**
 * Provides functionality to define an argument for a {@link FunctionDefinition}.
 */
export class ArgumentDefinition extends AbstractDefinition {

    /**
     * The type of this {@link ArgumentDefinition}
     */
    public type: C3Type

    public constructor(name: string,
                       type: C3Type,
                       decorators?: C3Decorator[],
                       metadata?: { [p: string]: any } | null) {
        super(name, decorators, metadata)
        this.type = type
    }
}

File: continuum-idl/src/main/java/org/kinotic/continuum/idl/api/schema/ArgumentDefinition.java
package org.kinotic.continuum.idl.api.schema;

import lombok.*;
import lombok.experimental.Accessors;

/**
 * Provides functionality to define an argument for a {@link FunctionDefinition}.
 * The context for equality here is the {@link FunctionDefinition}.
 * Given that no two arguments can have the same name in the same {@link FunctionDefinition}.
 * Created by navid on 2023-4-13
 */
@Getter
@Setter
@Accessors(chain = true)
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = false)
@ToString
public class ArgumentDefinition extends AbstractDefinition {

    /**
     * The name of this {@link ArgumentDefinition}, and the argument name for the {@link FunctionDefinition}
     */
    private String name;

    /**
     * This is the {@link C3Type} that defines the type of this argument.
     */
    @EqualsAndHashCode.Exclude // The context for equality here is the function definition
    private C3Type type;

}
